test/any/dao/api_service_dao-test.es6.js

```diff
@@ -0,0 +1,154 @@
+    }, ctx)
+        // Access delegate to trigger lazy factory responsible for validate().
+        .delegate).not.toThrow();
+  });
+
+  it('should pass baseURL to WorkerDAO', () => {
+    const baseURL = 'https://example.com/thingDAO';
+    let numWorkerDAOs = 0;
+    foam.CLASS({
+      name: 'MockWorkerDAO',
+      package: 'org.chromium.apis.web.test',
+      extends: 'org.chromium.apis.web.WorkerDAO',
+
+      properties: [
+        {
+          class: 'Int',
+          name: 'baseURLSetCount',
+        },
+        {
+          name: 'baseURL',
+          postSet: function(_, nu) {
+            expect(nu).toBe(baseURL);
+          },
+        },
+      ],
+
+      methods: [
+        function init() {
+          this.SUPER();
+          numWorkerDAOs++;
+          expect(this.baseURL).toBe(baseURL);
+        },
+      ],
+    });
+
+    ctx.register(pkg.test.MockWorkerDAO, 'org.chromium.apis.web.WorkerDAO');
+
+    pkg.ApiServiceDAO.create({
+      of: pkg.test.Thing,
+      baseURL,
+      workerRegistry: foam.box.Context.create(null, ctx).registry,
+    }, ctx)
+        // Access delegate to trigger lazy factory.
+        .delegate;
+
+    // Registry may involve instantiating more than one WorkerDAO, but at least
+    // one should be created.
+    expect(numWorkerDAOs).toBeGreaterThan(0);
+  });
+
+  it('should register worker DAO under given name', () => {
+    const serviceName = 'serviceName';
+    let numRegisters = 0;
+    foam.CLASS({
+      name: 'MockBoxRegistry',
+      package: 'org.chromium.apis.web.test',
+      extends: 'foam.box.BoxRegistryBox',
+
+      methods: [
+        function register(name, service, box) {
+          expect(name).toBe(serviceName);
+          numRegisters++;
+          return this.SUPER(name, service, box);
+        },
+      ],
+    });
+
+    const workerCtx = foam.createSubContext({});
+    workerCtx.register(pkg.test.MockBoxRegistry, 'foam.box.BoxRegistryBox');
+    const workerRegistry = foam.box.Context.create(null, workerCtx).registry;
+
+    pkg.ApiServiceDAO.create({
+      of: pkg.test.Thing,
+      name: serviceName,
+      baseURL: 'https://example.com/thingDAO',
+      workerRegistry,
+    }, ctx)
+        // Access delegate to trigger lazy factory.
+        .delegate;
+
+    expect(numRegisters).toBeGreaterThan(0);
+  });
+
+  it('should change delegate after async select()', done => {
+    const workerCtx = foam.createSubContext({});
+    const overwriteRegister = (cls, id) => {
+      ctx.register(cls, id);
+      workerCtx.register(cls, id);
+    }
+    overwriteRegister(foam.dao.NullDAO, 'org.chromium.apis.web.WorkerDAO');
+    overwriteRegister(foam.dao.NullDAO, 'foam.dao.RestDAO');
+
+    const dao = pkg.ApiServiceDAO.create({
+      of: pkg.test.Thing,
+      baseURL: 'https://example.com/thingDAO',
+      workerRegistry: foam.box.Context.create(null, workerCtx).registry,
+    }, ctx);
+    const initialDelegate = dao.delegate;
+    dao.delegate$.sub(done);
```

Comment:
Am I missing something, or is there nothing being asserted in this test?

Reply:
You're not missing anything. This test will time out if the `ApiServiceDAO` never changes its delegate after initialization. The code under test is the swapping out of the thread/process-local `RestDAO` for the now-ready `WorkerDAO`.

We could add assertions by reaching into the DAO and asserting that the delegate used to be a `WorkerDAO` and is now a `RestDAO`. I've been trying to keep the tests from depending too much on the implementation details though. WDYT? Should I add such assertions?